File: DSA/Backtracking/Code/NQueens.py
# N-Queens Problem using Backtracking in Python
N = 4
board = [[0]*N for _ in range(N)]

def is_safe(row, col):
    for i in range(col):
        if board[row][i]: return False
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j]: return False
    for i, j in zip(range(row, N, 1), range(col, -1, -1)):
        if board[i][j]: return False
    return True

def print_board():
    for i in range(N):
        print(' '.join('Q' if board[i][j] else '.' for j in range(N)))
    print()

def solve_nq_util(col):
    if col == N:
        print_board()
        return
    for i in range(N):
        if is_safe(i, col):
            board[i][col] = 1
            solve_nq_util(col+1)
            board[i][col] = 0

print("=== N-Queens Problem (Python) ===\n")
solve_nq_util(0)
print("Time Complexity: O(N!)")
print("Space Complexity: O(N^2)")

File: DSA/Backtracking/Code/NQueens.c
#include <stdio.h>
#include <math.h>

// N-Queens Problem using Backtracking
#define N 4

int board[N][N];

int isSafe(int row, int col) {
    for (int i = 0; i < col; i++)
        if (board[row][i]) return 0;
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--)
        if (board[i][j]) return 0;
    for (int i = row, j = col; i < N && j >= 0; i++, j--)
        if (board[i][j]) return 0;
    return 1;
}

void printBoard() {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++)
            printf("%c ", board[i][j] ? 'Q' : '.');
        printf("\n");
    }
    printf("\n");
}

void solveNQUtil(int col) {
    if (col == N) {
        printBoard();
        return;
    }
    for (int i = 0; i < N; i++) {
        if (isSafe(i, col)) {
            board[i][col] = 1;
            solveNQUtil(col+1);
            board[i][col] = 0;
        }
    }
}

int main() {
    printf("=== N-Queens Problem (C) ===\n\n");
    solveNQUtil(0);
    printf("Time Complexity: O(N!)\nSpace Complexity: O(N^2)\n");
    return 0;
}
